Finalize generation script

thepetk opened the pull request:
### What does this PR do?:

The PR finalizes the automatic conversion of the shared gitops resources between the `ai-lab-app` and the `ai-lab-helm-chart`.

More specifically it introduces the following updates:

* A script called `generate.sh` is added and is the main point to update all the shared resources. Along with the script a workflow has been added to check in every new PR that we are aligned with the `ai-lab-app` repo. This way we guarantee that both `ai-lab-app` & `ai-lab-samples` are always up-to-date. Finally, a quick readme has been added to capture all this information. I find the best location to be the `ai-software-templates` directory which contains all the related charts.
* In regards to the `chatbot-ai-sample` chart, all the updates brought from https://github.com/redhat-ai-dev/ai-lab-helm-charts/pull/31 are now tested and captured in the readme tmpl (and ofc in the readme) of the chart.
* Finally, the `values.dbRequired` is removed for now from the new resource content as it is only related with the RAG case. I think the best approach here is to create a new chart that will cover this case and not use the chatbot chart for this.

Apart from the above, also some other details are:
* To make sure that the `existingModelServer` will remove the `model-server` deployment if set, the conversion script now adds on the top and bottom of all `*model-server*` files a condition.
* As you can notice some updates are also pulled from ai-lab-app after the formatting [PR](https://github.com/redhat-ai-dev/ai-lab-app/pull/41) that was merged recently.

### Which issue(s) this PR fixes:

Fixes RHDHPAI-379

### PR acceptance criteria:

Testing and documentation do not need to be complete in order for this PR to be approved. We just need to ensure tracking issues are opened and linked to this PR, if they are not in the PR scope due to various constraints.

- [ ] Tested and Verified

<!-- _I have verified that the changes were tested manually_ -->

- [ ] Documentation (READMEs, Product Docs, Blogs, Education Modules, etc.)

 <!-- _This includes READMEs, Product Docs, Blogs, Education Modules, etc._ -->

### How to test changes / Special notes to the reviewer:

In regards to testing, I've tested the following cases:

* `vLLM` support. If `vllmSelected` is set the model server deployed is based on this technology. To test this case you need to provision a cluster with GPU support:

![image](https://github.com/user-attachments/assets/89ca0bdb-5ba4-4e09-ba58-152e0d7e74b7)

* The existing model server case, along with the bearer authentication support. This skips the model-server deployment and service creation:

![existing-model-server-vs-vllm](https://github.com/user-attachments/assets/de400481-7036-4823-858e-f82fc4f7d83d)

* A test run for the new workflow is [here](https://github.com/thepetk/ai-lab-helm-charts/actions/runs/12994160859/job/36238015683)

thepetk committed:
Remove support for dbRequired

Signed-off-by: thepetk <thepetk@gmail.com>

thepetk committed:
Pull updates after gitops lint

Signed-off-by: thepetk <thepetk@gmail.com>

thepetk committed:
Add newly imported values

Signed-off-by: thepetk <thepetk@gmail.com>

thepetk committed:
Update values

Signed-off-by: thepetk <thepetk@gmail.com>

thepetk committed:
Fix defaults

Signed-off-by: thepetk <thepetk@gmail.com>

thepetk committed:
Add modelMaxLength value

Signed-off-by: thepetk <thepetk@gmail.com>

thepetk committed:
Update resources after testing

Signed-off-by: thepetk <thepetk@gmail.com>

thepetk committed:
Update tmpl json schema

Signed-off-by: thepetk <thepetk@gmail.com>

thepetk committed:
Update readme template

Signed-off-by: thepetk <thepetk@gmail.com>

thepetk committed:
Add readme for generate.sh

Signed-off-by: thepetk <thepetk@gmail.com>

thepetk committed:
Add the generate.sh

Signed-off-by: thepetk <thepetk@gmail.com>

thepetk committed:
Rename github action

Signed-off-by: thepetk <thepetk@gmail.com>

thepetk commented:
WIP: because I'm working on an issue with the MODEL_ENDPOINT_BEARER env var and the update of the deployment done by the tekton pipeline

thepetk committed:
Bump chatbot version

Signed-off-by: thepetk <thepetk@gmail.com>

gabemontero commented on charts/ai-software-templates/README.md:
```diff
@@ -0,0 +1,9 @@
+## ai-software-templates
+
+Apart from the [pipeline-install](./pipeline-install/) and the [pipeline-setup](./pipeline-setup/) which are tools to help you install the tekton pipelines along with your charts, the rest of charts under this directory are based on the [redhat-ai-dev/ai-lab-app](https://github.com/redhat-ai-dev/ai-lab-app) gitops resources repo.
```
```suggestion
Apart from the [pipeline-install](./pipeline-install/) and the [pipeline-setup](./pipeline-setup/) which are tools to help you install the Tekton Pipelines along with your charts, the rest of charts under this directory are based on the [redhat-ai-dev/ai-lab-app](https://github.com/redhat-ai-dev/ai-lab-app) gitops resources repo.
```

gabemontero commented on charts/ai-software-templates/README.md:
```diff
@@ -0,0 +1,9 @@
+
+### Pull automatically latest changes
+
+To pull all latest changes for the charts based on `ai-lab-app` you can simply run the `generate.sh` script placed in the root dir of this repository.
```
```suggestion
To pull all the latest changes for the charts from the `ai-lab-app` repository you can simply run the `generate.sh` script placed in the root dir of this repository.
```

gabemontero commented on charts/ai-software-templates/README.md:
```diff
@@ -0,0 +1,9 @@
+
+To pull all latest changes for the charts based on `ai-lab-app` you can simply run the `generate.sh` script placed in the root dir of this repository.
+
+The script will clone the `ai-lab-app` and will convert the necessary resources to helm chart template files. Each chart has a corresponding `env` file under `scripts/envs` dir which helps us configure the behavior of the conversion process.
```
```suggestion
The script will clone the `ai-lab-app` and will convert the necessary resources to Helm chart template files. Each chart has a corresponding `env` file under the `scripts/envs` dir which helps us configure the behavior of the conversion process.
```

gabemontero commented on charts/ai-software-templates/chatbot/README.md:
```diff
@@ -10,7 +10,12 @@ This Helm Chart deploys a Large Language Model (LLM)-enabled [chat bot applicati
 
 ### The Model Service
-Based on the `llama.cpp` inference server and related to the [ai-lab-recipes model server](https://github.com/containers/ai-lab-recipes/tree/main/model_servers/llamacpp_python).
+By default the `chatbot-ai-sample` supports the `llama.cpp` inference server and related to the [ai-lab-recipes model server](https://github.com/containers/ai-lab-recipes/tree/main/model_servers/llamacpp_python).
```
```suggestion
By default the `chatbot-ai-sample` uses the `llama.cpp` inference via the [ai-lab-recipes model server](https://github.com/containers/ai-lab-recipes/tree/main/model_servers/llamacpp_python).
```

thepetk replied:
Fixed in [87d9c1e](https://github.com/redhat-ai-dev/ai-lab-helm-charts/pull/33/commits/87d9c1eb6630c67aa40b4c29ed17014d321a90d5)

gabemontero commented on charts/ai-software-templates/chatbot/README.md:
```diff
@@ -10,7 +10,12 @@ This Helm Chart deploys a Large Language Model (LLM)-enabled [chat bot applicati
-Based on the `llama.cpp` inference server and related to the [ai-lab-recipes model server](https://github.com/containers/ai-lab-recipes/tree/main/model_servers/llamacpp_python).
+By default the `chatbot-ai-sample` supports the `llama.cpp` inference server and related to the [ai-lab-recipes model server](https://github.com/containers/ai-lab-recipes/tree/main/model_servers/llamacpp_python).
+
+However, the usage of `vLLM` model services or existing model services is also supported:
```
```suggestion
However, the use of `vLLM` to deploy model services, as well as the use of existing model services is also supported:
```

thepetk replied:
Fixed in [87d9c1e](https://github.com/redhat-ai-dev/ai-lab-helm-charts/pull/33/commits/87d9c1eb6630c67aa40b4c29ed17014d321a90d5)

gabemontero commented on charts/ai-software-templates/chatbot/README.md.gotmpl:
```diff
@@ -10,7 +10,12 @@
 
 ### The Model Service
-Based on the `llama.cpp` inference server and related to the [ai-lab-recipes model server](https://github.com/containers/ai-lab-recipes/tree/main/model_servers/llamacpp_python).
+By default the `chatbot-ai-sample` supports the `llama.cpp` inference server and related to the [ai-lab-recipes model server](https://github.com/containers/ai-lab-recipes/tree/main/model_servers/llamacpp_python).
```
can you replicate my suggestions for this from the README.md file

thepetk replied:
Fixed in [87d9c1e](https://github.com/redhat-ai-dev/ai-lab-helm-charts/pull/33/commits/87d9c1eb6630c67aa40b4c29ed17014d321a90d5)

gabemontero commented on charts/ai-software-templates/chatbot/README.md.gotmpl:
```diff
@@ -10,7 +10,12 @@
-Based on the `llama.cpp` inference server and related to the [ai-lab-recipes model server](https://github.com/containers/ai-lab-recipes/tree/main/model_servers/llamacpp_python).
+By default the `chatbot-ai-sample` supports the `llama.cpp` inference server and related to the [ai-lab-recipes model server](https://github.com/containers/ai-lab-recipes/tree/main/model_servers/llamacpp_python).
+
+However, the usage of `vLLM` model services or existing model services is also supported:
```
can you replicate my suggestions for this form the REAMDE.md file

thepetk replied:
Fixed in [87d9c1e](https://github.com/redhat-ai-dev/ai-lab-helm-charts/pull/33/commits/87d9c1eb6630c67aa40b4c29ed17014d321a90d5)

gabemontero reviewed (changes requested):
Just some review of the README's with some minor wording suggestions

I'm booting up a cluster now and will checkout this branch and given a go later today @thepetk - thanks

gabemontero commented on scripts/convert-gitops-template.sh:
```diff
@@ -33,6 +33,7 @@ convert() {
     replace "values\.modelEndpointSecretKey" ".Values.model.modelEndpointSecretKey" "$input_file"
     replace "values\.existingModelServer" ".Values.model.existingModelServer" "$input_file"
     replace "values\.dbRequired" ".Values.model.dbRequired" "$input_file"
+    replace "values\.maxModelLength" ".Values.model.maxModelLength" "$input_file"
```
at some point can you confirm if you ran `shellcheck` against this file for the changes here?

thepetk replied:
I think I'll add today one or two changes for the bearer auth case and then I'll also run shellcheck

thepetk replied:
I ran the shellcheck and I added a minor suggestion. (see 
99d8f42) 
I'll resolve this one!

gabemontero commented:
Testing the 3 charts worked for me @thepetk , including an update to the app getting rolled out.  I'm guessing until we have an update to `ai-lab-app` or `ai-lab-samples` the `generate.sh` script won't produce any changes (it did not for me when I ran locally).

thepetk commented:
> Testing the 3 charts worked for me @thepetk , including an update to the app getting rolled out. I'm guessing until we have an update to `ai-lab-app` or `ai-lab-samples` the `generate.sh` script won't produce any changes (it did not for me when I ran locally).

Yeah it should be this way and pretty much this is the main reason I added the workflow just to be sure every time we are up-to-date with the ai-lab-app dependency. For the `ai-lab-samples` as we directly populating the application repo from the `ai-lab-samples` we are automatically up-to-date.

thepetk committed:
Update charts/ai-software-templates/README.md

Co-authored-by: Gabe Montero <gmontero@redhat.com>

thepetk committed:
Update charts/ai-software-templates/README.md

Co-authored-by: Gabe Montero <gmontero@redhat.com>

thepetk committed:
Update charts/ai-software-templates/README.md

Co-authored-by: Gabe Montero <gmontero@redhat.com>

thepetk committed:
Update readme template

gabemontero commented:
> > Testing the 3 charts worked for me @thepetk , including an update to the app getting rolled out. I'm guessing until we have an update to `ai-lab-app` or `ai-lab-samples` the `generate.sh` script won't produce any changes (it did not for me when I ran locally).
> 
> Yeah it should be this way and pretty much this is the main reason I added the workflow just to be sure every time we are up-to-date with the ai-lab-app dependency. For the `ai-lab-samples` dependency, I think we are ok, as we directly populating the application repo from the `ai-lab-samples` we are automatically up-to-date. WDYT?

Sure all that is good.  I'm just saying we can't fully validate until a change is made to those repo(s).

Before you move your Jira to closed, were you planning on making some sort of change in those repo(s) that would get caught by your github workflow, or get seen if someone ran `generate.sh` locally?

thepetk commented:
> > > Testing the 3 charts worked for me @thepetk , including an update to the app getting rolled out. I'm guessing until we have an update to `ai-lab-app` or `ai-lab-samples` the `generate.sh` script won't produce any changes (it did not for me when I ran locally).
> > 
> > 
> > Yeah it should be this way and pretty much this is the main reason I added the workflow just to be sure every time we are up-to-date with the ai-lab-app dependency. For the `ai-lab-samples` dependency, I think we are ok, as we directly populating the application repo from the `ai-lab-samples` we are automatically up-to-date. WDYT?
> 
> Sure all that is good. I'm just saying we can't fully validate until a change is made to those repo(s).
> 
> Before you move your Jira to closed, were you planning on making some sort of change in those repo(s) that would get caught by your github workflow, or get seen if someone ran `generate.sh` locally?

I think we are getting new changes right? Because we are importing the changes from the linter merged in the `ai-lab-app` recently. So the script ran and imported the updates on the ai-lab-app files. This can be verified by the workflow now as now that we run it again it doesn't introduce anything new.

thepetk committed:
Make app-config permanent

thepetk commented:
Example screenshot using existing model server, bearer auth and the updated pipeline introduced by https://github.com/redhat-ai-dev/rhdh-pipelines/pull/11

![Screenshot from 2025-01-29 15-00-38](https://github.com/user-attachments/assets/fbb2dd67-94be-4f29-ad55-2605e31d4e87)

Btw the whole experience with an existing model server is way faster, the application is up and running very fast, so the git repo and the pipelines.

gabemontero commented:
> > > > Testing the 3 charts worked for me @thepetk , including an update to the app getting rolled out. I'm guessing until we have an update to `ai-lab-app` or `ai-lab-samples` the `generate.sh` script won't produce any changes (it did not for me when I ran locally).
> > > 
> > > 
> > > Yeah it should be this way and pretty much this is the main reason I added the workflow just to be sure every time we are up-to-date with the ai-lab-app dependency. For the `ai-lab-samples` dependency, I think we are ok, as we directly populating the application repo from the `ai-lab-samples` we are automatically up-to-date. WDYT?
> > 
> > 
> > Sure all that is good. I'm just saying we can't fully validate until a change is made to those repo(s).
> > Before you move your Jira to closed, were you planning on making some sort of change in those repo(s) that would get caught by your github workflow, or get seen if someone ran `generate.sh` locally?
> 
> I think we are getting new changes right? Because we are importing the changes from the linter merged in the `ai-lab-app` recently. So the script ran and imported the updates on the ai-lab-app files. This can be verified by the workflow now as now that we run it again it doesn't introduce anything new.

Either I'm mis-interpreting what you are saying, or I'm not making myself sufficiently clear, or something along those lines @thepetk 

Let me try this way:  At some point we should be able to submit at PR against this repo, and see the print at  https://github.com/redhat-ai-dev/ai-lab-helm-charts/pull/33/files#diff-119dd431be9907b10ad9714895fb89d998ccd55bf25f3cc6bb4f1573ed7ca7e2R21 occur because on of those other repositories have been updated, correct?

Ideally we test that flow with perhaps a simple change, non effecting change in one of the other repos, merge in one of those repos, and then see your new workflow flag that `generate.sh` needs to be executed.

Correct?

thepetk commented:
> > > > > Testing the 3 charts worked for me @thepetk , including an update to the app getting rolled out. I'm guessing until we have an update to `ai-lab-app` or `ai-lab-samples` the `generate.sh` script won't produce any changes (it did not for me when I ran locally).
> > > > 
> > > > 
> > > > Yeah it should be this way and pretty much this is the main reason I added the workflow just to be sure every time we are up-to-date with the ai-lab-app dependency. For the `ai-lab-samples` dependency, I think we are ok, as we directly populating the application repo from the `ai-lab-samples` we are automatically up-to-date. WDYT?
> > > 
> > > 
> > > Sure all that is good. I'm just saying we can't fully validate until a change is made to those repo(s).
> > > Before you move your Jira to closed, were you planning on making some sort of change in those repo(s) that would get caught by your github workflow, or get seen if someone ran `generate.sh` locally?
> > 
> > 
> > I think we are getting new changes right? Because we are importing the changes from the linter merged in the `ai-lab-app` recently. So the script ran and imported the updates on the ai-lab-app files. This can be verified by the workflow now as now that we run it again it doesn't introduce anything new.
> 
> Either I'm mis-interpreting what you are saying, or I'm not making myself sufficiently clear, or something along those lines @thepetk
> 
> Let me try this way: At some point we should be able to submit at PR against this repo, and see the print at https://github.com/redhat-ai-dev/ai-lab-helm-charts/pull/33/files#diff-119dd431be9907b10ad9714895fb89d998ccd55bf25f3cc6bb4f1573ed7ca7e2R21 occur because on of those other repositories have been updated, correct?
> 
> Ideally we test that flow with perhaps a simple change, non effecting change in one of the other repos, merge in one of those repos, and then see your new workflow flag that `generate.sh` needs to be executed.
> 
> Correct?

Yeap! That's correct! Also apologies if my message was not clear.

I meant that in the current PR we can confirm that the `generate.sh` works as the PR imports changes merged in `ai-lab-app` from [this commit](https://github.com/redhat-ai-dev/ai-lab-app/commit/7f6422f482ca82eed29fe79448d153aa704773b6). Those updates are imported in the `ai-lab-helm-charts` repo as the last time we imported changes was before this commit (in this [PR](https://github.com/redhat-ai-dev/ai-lab-helm-charts/pull/31)).  This can be confirmed from the [corresponding workflow](https://github.com/redhat-ai-dev/ai-lab-helm-charts/actions/runs/13033973651/job/36359688607?pr=33).

No to test also the oposite, e.g that the workflow fails when "no-change-was-introduced", I've added [a quick change](https://github.com/thepetk/ai-lab-app/commit/7f363da99b295a2a104dc5f949d5b94daf44e513) in my fork of the `ai-lab-app`.

In my `ai-lab-helm-charts` fork, I've updated the script to point to my `ai-lab-app` fork and opened a test PR. I can confirm the workflow fails (status check failed [here](https://github.com/thepetk/ai-lab-helm-charts/actions/runs/13052515220/job/36415773020?pr=2).

Let me know if I'm miss

gabemontero commented:
> > WIP: because I'm working on an issue with the MODEL_ENDPOINT_BEARER env var and the update of the deployment done by the tekton pipeline
> 
> I've managed to make the bearer auth work. In order to achieve that I've opened a [PR in the rhdh-pipelines](https://github.com/redhat-ai-dev/rhdh-pipelines/pull/11) side and I made an update to the app-config so it can stay permanently as a configmap so we can fetch each time the necessary variables.
> 
> @gabemontero PTAL :)

So is there a situation where we need to merge the rhdh-pipelines change *BEFORE* this one, or are you just citing the relationship, but they can merge independently @thepetk ?

thepetk commented:
> > > WIP: because I'm working on an issue with the MODEL_ENDPOINT_BEARER env var and the update of the deployment done by the tekton pipeline
> > 
> > 
> > I've managed to make the bearer auth work. In order to achieve that I've opened a [PR in the rhdh-pipelines](https://github.com/redhat-ai-dev/rhdh-pipelines/pull/11) side and I made an update to the app-config so it can stay permanently as a configmap so we can fetch each time the necessary variables.
> > @gabemontero PTAL :)
> 
> So is there a situation where we need to merge the rhdh-pipelines change _BEFORE_ this one, or are you just citing the relationship, but they can merge independently @thepetk ?

I'm quite on the fence, but honestly I don't see a blocker but as you say there's a relationship.

My plan is:
* Merge this one, and the tekton will not be affected.
* Merge the update on tekton so we can cover the case of the bearer auth.
* Finally, I'll release a new helm chart version on the openshift-charts so we can have all this functionality there too.

thepetk commented:
> If that occurred, I am unable to determine that. Is it possibly one of the 18 commits that pulled in that change that you can point me to?

I think that one is a good example: https://github.com/redhat-ai-dev/ai-lab-helm-charts/pull/33/commits/ad618d8bde11b1df022bc3ddda0fb609fafc57f3
